### ISSUE TITLE
AWS ALB and Kafka rest proxy

### ISSUE THREAD

**Issue by nkarw** [confluentinc/kafka-rest#684]
Currently, we use an AWS ALB to route incoming traffic to an EC2 instance, which in turn adds the data to our Kafka brokers. Can we replace this EC2 with the Kafka rest proxy so that the ALB gets an HTTP(S) end point to direct the traffic to the kafka brokers.
Does the kafka rest proxy support this feature?

Thanks

**Comment by cedricve** [comment]
looking for something similar..

**Comment by mithunmanohar** [comment]
Yes. It should be possible. I have tested two patterns for our REST proxy implementation and found it working.
1. R53 > ALB( listner configured on port 443, using AWS ACM https certificate) > Target Group which forwards it EC2/ ECS instance id (running REST proxy on any port)

AWS API gateway > VPC integration > VPC link > NLB > Listner on port 443 > Target Group > forward it EC2/ ECS instance id (running REST proxy on any port)

**Comment by 0ni0nrings** [comment]
> Yes. It should be possible. I have tested two patterns for our REST proxy implementation and found it working.
> 
>     1. R53 > ALB( listner configured on port 443, using AWS ACM https certificate) > Target Group which forwards it EC2/ ECS instance id (running REST proxy on any port)
> 
>     2. AWS API gateway > VPC integration > VPC link > NLB > Listener on port 443 > Target Group > forward it EC2/ ECS instance id (running REST proxy on any port)
>        Note that here the HTTPS traffic terminates at ALB and traffic is routed to REST proxy using HTTP

Hi @mithunmanohar you have two interesting use cases. Please, can I ask you to elaborate on them? Or if you could point me to some reading material that will help me in testing these hypothesis.

**Comment by mithunmanohar** [comment]
Hi @0ni0nrings ,
1. You can run rest-proxy stand-alone using its Docker container (https://hub.docker.com/r/confluentinc/cp-kafka-rest). As mentioned above I deployed this container in an ECS task. It can be run on an Ec2 as well.
2. I don't have any exact document on the implementation. But it is similar to https://medium.com/swlh/aws-api-gateway-private-integration-with-http-api-and-a-vpc-link-602360a1cd84 . Instead of Ec2 in this case I had traffic going to ECS task running rest proxy container
3. You should be able to register Ec2 instance as a target in load balancer using instance ID or IP

**Comment by 0ni0nrings** [comment]
thanks, Murali for replying back.
I will try to first get it right on ec2 instance & then move to a ecs/fargate. Right now I am having trouble getting traffic from load balancer to ec2 instances. I am getting 502 error when I curl the dns address of load balancer. Did you have similar trouble?
```
curl -X GET "http://rest-proxy-lb-1694149832.us-east-1.elb.amazonaws.com/v3/clusters/"
Note: Unnecessary use of -X or --request, GET is already inferred.
*   Trying 52.22.149.252...
* TCP_NODELAY set
* Connected to rest-proxy-lb-1694149832.us-east-1.elb.amazonaws.com (52.22.149.252) port 80 (#0)
> GET /v3/clusters/ HTTP/1.1
> Host: rest-proxy-lb-1694149832.us-east-1.elb.amazonaws.com
> User-Agent: curl/7.64.1
> Accept: */*
>
< HTTP/1.1 502 Bad Gateway
< Server: awselb/2.0
< Date: Mon, 22 Mar 2021 02:15:51 GMT
< Content-Type: text/html
< Content-Length: 122
< Connection: keep-alive
<
<html>
<head><title>502 Bad Gateway</title></head>
<body>
<center><h1>502 Bad Gateway</h1></center>
</body>
</html>
* Connection #0 to host rest-proxy-lb-1694149832.us-east-1.elb.amazonaws.com left intact
* Closing connection 0
```

________________________________

**Comment by mithunmanohar** [comment]
@0ni0nrings  This looks like something specific to your configuration and AWS environment. What is the listener rule set up in ALB and to which target group the traffic is forwarded? 
Is the target group to which the traffic is forwarded, is 'Registered targets' health check passing and showing the ec2 instance as healthy? As long as the health check fails, the traffic is not forwarded to the intended instance from the ALB. If the health check is failing - make sure the security group of EC2 allows traffic on the port configured. 
If you have an AWS premium organization support plan - raise a ticket with them and they can quickly guide you on this.

**Comment by DawoodOballi** [comment]
How can I configure the ALB to pass health checks given there isnt a health enpoint for the rest proxy?
